src/storage/src/compaction/task.rs

```diff
@@ -12,23 +12,126 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::collections::HashSet;
+use std::sync::{Arc, Mutex};
+
+use common_telemetry::{error, info};
+use object_store::ObjectStore;
+use store_api::logstore::LogStore;
+use uuid::Uuid;
+
+use crate::compaction::writer::build_sst_reader;
 use crate::error::Result;
-use crate::sst::{FileHandle, Level};
+use crate::manifest::action::RegionEdit;
+use crate::manifest::region::RegionManifest;
+use crate::region::{RegionWriterRef, SharedDataRef};
+use crate::schema::RegionSchemaRef;
+use crate::sst::parquet::{ParquetWriter, Source};
+use crate::sst::{AccessLayerRef, FileHandle, FileMeta, Level, SstInfo, WriteOptions};
+use crate::wal::Wal;
 
 #[async_trait::async_trait]
 pub trait CompactionTask: Send + Sync + 'static {
-    async fn run(&self) -> Result<()>;
+    async fn run(self) -> Result<()>;
 }
 
 #[allow(unused)]
-pub(crate) struct CompactionTaskImpl {
-    inputs: Vec<CompactionInput>,
+pub(crate) struct CompactionTaskImpl<S: LogStore> {
+    pub schema: RegionSchemaRef,
+    pub sst_layer: AccessLayerRef,
+    pub outputs: Vec<CompactionOutput>,
+    pub writer: RegionWriterRef,
+    pub shared_data: SharedDataRef,
+    pub wal: Wal<S>,
+    pub manifest: RegionManifest,
+}
+
+impl<S: LogStore> CompactionTaskImpl<S> {
+    /// Compacts inputs SSTs, returns `(output file, compacted input file)`.
+    async fn merge_ssts(&mut self) -> Result<(Vec<FileMeta>, Vec<FileMeta>)> {
+        let mut futs = Vec::with_capacity(self.outputs.len());
+        let compacted_inputs = Arc::new(Mutex::new(HashSet::new()));
+
+        for output in self.outputs.drain(..) {
+            let schema = self.schema.clone();
+            let sst_layer = self.sst_layer.clone();
+            let object_store = self.sst_layer.object_store();
+            let compacted = compacted_inputs.clone();
```

Comment:
Could we just extend the `compacted_inputs` before creating the future? As we only return `Ok` when all futures are done, it isn't necessary to collect compacted_inputs inside the async tasks.
```rust
let compacted_inputs = Vec::new();
for output in self.outputs.drain(..) {
    compacted_inputs.extend(output.inputs.iter().map(|f| xxx);
}
```

Reply:
> Could we just extend the `compacted_inputs` before creating the future? As we only return `Ok` when all futures are done, it isn't necessary to collect compacted_inputs inside the async tasks.
> 
> ```rust
> let compacted_inputs = Vec::new();
> for output in self.outputs.drain(..) {
>     compacted_inputs.extend(output.inputs.iter().map(|f| xxx);
> }
> ```

True, I intended to allow partial success, but looks like it's not necessary.